### PR TITLE
[travis] generate size report

### PR DESCRIPTION
**Pull request by bukepo**
This PR split the jobs into 2 stages: lint, test. In the lint stage, generate size diff report and post into the pull request.

### CODE REVIEW

File: .travis/size-report [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+|  | #${TRAVIS_PULL_REQUEST} | ${new[0]} | ${new[1]} | ${new[2]} | ${new[3]} |
+|  | +/- | $((${new[0]} - ${old[0]})) | $((${new[1]} - ${old[1]})) | $((${new[2]} - ${old[2]})) | $((${new[3]} - ${old[3]})) |
+EOF
+}
+
+size_report_post()
+{
+    [[ "${TRAVIS_PULL_REQUEST}" != false ]] || return 0
+
+    curl -H "Authorization: token ${GITHUB_TOKEN}" -X POST \
```
Comment (jwhui) [inline .travis/size-report]: Does this work for the fork-and-pull model? [Travis CI docs](https://docs.travis-ci.com/user/environment-variables/) states "we do not provide these values to untrusted builds, triggered by pull requests from another repository."

Comment (bukepo) [reply .travis/size-report]: I didn't noticed that. I will create a GitHub app for this. Thoughts?

Comment (bukepo) [reply .travis/size-report]: I didn't noticed that. I will create a GitHub app for this. Thoughts?

Comment (jwhui) [reply .travis/size-report]: I think a GitHub app should work.

**Comment by bukepo** [comment]
The size report is working!

**Comment by abtink** [comment]
Cool.

File: src/core/mac/mac.hpp [hunk 2]
```diff
@@ -683,6 +683,8 @@ class Mac : public InstanceLocator
 
     static const char *OperationToString(Operation aOperation);
 
+    SubMac mSubMac;
```
Comment (abtink) [inline src/core/mac/mac.hpp]: I guess you changed this for testing only. Right? (and will remove it before merge?)
I expect this change can cause a RAM size increase (by ~8 or so ). But it's interesting that we see code size change as well...

File: .travis/script.sh [hunk 3]
```diff
@@ -450,6 +450,46 @@ build_samr21() {
...
+    OPENTHREAD_FLAGS="BORDER_AGENT=1 BORDER_ROUTER=1 CHANNEL_MANAGER=1 CHANNEL_MONITOR=1 CHILD_SUPERVISION=1 COAP=1 COAPS=1 COMMISSIONER=1 DHCP6_CLIENT=1 DHCP6_SERVER=1 DIAGNOSTIC=1 DISABLE_DOC=1 DNS_CLIENT=1 ECDSA=1 FULL_LOGS=1 JAM_DETECTION=1 JOINER=1 LINK_RAW=1 MAC_FILTER=1 MTD_NETDIAG=1 SERVICE=1 SLAAC=1 SNTP_CLIENT=1 TIME_SYNC=1 UDP_FORWARD=1"
+
+    git checkout -- . || die
+    git clean -xfd || die
+    ./bootstrap || die
+    make -f examples/Makefile-nrf52840 ${OPENTHREAD_FLAGS} || die
+    mv output/nrf52840 ../output/nrf52840-b
+
+    git checkout ${TRAVIS_BRANCH}
+    git submodule update --init
```
Comment (abtink) [inline .travis/script.sh]: Just curios:  How are we getting to master branch for comparison? 
Is it somehow derived from the PR branch itself (or is it top of master?)

Comment (bukepo) [reply .travis/script.sh]: ${TRAVIS_BRANCH} will be master for pull requests. We just switch to master branch here.

File: src/core/mac/mac.hpp (hunk 2 above)
Comment (bukepo) [reply src/core/mac/mac.hpp]: Yes, I will revert this change, which just shows this works.

File: .travis/script.sh (hunk 3 above)
Comment (abtink) [reply .travis/script.sh]: Perfect. Thanks.

**Comment by abtink** [comment]
Thanks @bukepo. This is great and super helpful...  :)

**Comment by bukepo** [comment]
@abtink just reverted the change to `Mac::mSubMac`.

File: .travis/script.sh (hunk 3 above)
Comment (jwhui) [reply .travis/script.sh]: Is `${TRAVIS_BRANCH}` the current HEAD of the target branch? Or the common ancestor (i.e. git-merge-base)? I think the common ancestor is what we want.

Comment (bukepo) [reply .travis/script.sh]: Is HEAD of master better? I think we want to know the size diff before and after the merge, right?

Comment (bukepo) [reply .travis/script.sh]: @jwhui updated. https://github.com/openthread/size-report/pull/1

**Comment by jwhui** [comment]
@bukepo, couple questions:

1. Can we have size-report update an existing comment if one exists, rather than posting a new comment on each run?

1. Can we include the base commit id in the comment somewhere rather than just embedding in the link?

**Comment by bukepo** [comment]
Hi @jwhui , I think both are doable. However, is posting multiple comments more user friendlyIf a pull request has many comments, if we only update an existing size-report, it could be very far away from the latest comment, which can easily be ignored. I also see codecov posts multiple reports.
For the base commit id, I can add it now.

**Comment by jwhui** [comment]
[Codecov docs](https://docs.codecov.io/docs/pull-request-comments#section-behavior) state default is "update, if exists. Otherwise post new". The fact that it posts multiple comments is a bug they are still trying to resolve. See the "edited" pull down on the first Codecov comment on this. You'll find most of the revisions in there.

I think having multiple size reports (one for each push) can make it difficult to view the comments themselves since it causes more scrolling.

**Comment by bukepo** [comment]
OK, I will update to edit existing size report.

**Comment by bukepo** [comment]
@jwhui updated to edit existing report.

**Comment by bukepo** [comment]
@jwhui resolved conflicts.

**Comment by bukepo** [comment]
@jwhui any more concerns for this PR?

File: .travis.yml [hunk 4]
```diff
@@ -158,16 +162,44 @@ matrix:
...
+            - ubuntu-toolchain-r-test
+          packages:
+            - clang-format-6.0
+      script: .travis/script.sh
+    - env: BUILD_TARGET="size-report"
+      os: linux
+      compiler: gcc
+      script: .travis/script.sh VERBOSE=1
+      if: type = pull_request
+stages:
```
Comment (jwhui) [inline .travis.yml]: nit: prefer to put `stages` above `jobs` since the former groups the latter.

Comment (bukepo) [reply .travis.yml]: moved above jobs.

**Pull request merged**